### PR TITLE
Don't keep hard references to Table

### LINKED ISSUE #2130: Kafka subscription / table not cleaned up after deleting table
It appears that deleting a kafka table does not free the resources of the underlying kafka subscription or table. Demonstrated in python (will try in groovy soon).

### PR DESCRIPTION
**Pull request by devinrsmith**
Fixes #2130

Co-authored-by: Ryan Caudy <ryan@deephaven.io>

### CODE REVIEW

File: server/src/main/java/io/deephaven/server/appmode/ApplicationServiceGrpcImpl.java [hunk 1]
```diff
@@ -67,149 +67,264 @@ public ApplicationServiceGrpcImpl(final AppMode mode,
...
 
-            // let's not duplicate information when a client does actually join
-            addedFields.clear();
-            removedFields.clear();
-            updatedFields.clear();
+        boolean removed = false;
+
+        @Override
+        CUR out() {
+            // If an object does initially exists, there are two outcomes:
```
Comment (rcaudy) [inline server/src/main/java/io/deephaven/server/appmode/ApplicationServiceGrpcImpl.java:205]: IntelliJ is complaining that "exists" is plural.

File: server/src/main/java/io/deephaven/server/appmode/ApplicationServiceGrpcImpl.java [hunk 2]
```diff
@@ -67,149 +67,264 @@ public ApplicationServiceGrpcImpl(final AppMode mode,
...
     }
 
     @Override
-    public synchronized void onNewField(final ApplicationState app, final Field<?> field) {
+    public synchronized void onRemoveField(ApplicationState app, Field<?> oldField) {
         if (!mode.hasVisibilityToAppExports()) {
             return;
         }
+        final AppFieldId id = AppFieldId.from(app, oldField.name());
+        // todo: or should we rely on the known type
```
Comment (rcaudy) [inline server/src/main/java/io/deephaven/server/appmode/ApplicationServiceGrpcImpl.java:135]: I think that is "safest". Same for description. Might be annoying given the current code pattern, though.

File: server/src/main/java/io/deephaven/server/appmode/ApplicationServiceGrpcImpl.java [hunk 3]
```diff
@@ -304,7 +378,7 @@ public void run() {
         // must be sync wrt parent
         private void markRunning() {
             if (!isScheduled) {
-                throw new IllegalStateException("Job is running without being scheduled");
+                // throw new IllegalStateException("Job is running without being scheduled");
```
Comment (rcaudy) [inline server/src/main/java/io/deephaven/server/appmode/ApplicationServiceGrpcImpl.java:381]: Why did you need to comment this?

**Review by rcaudy — COMMENTED** [review]
Seems like a lot of code for the state machines, etc. That said, I think it does a pretty good job of cleaning up the reachability and liveness issues, and it's overall simpler/more obviously correct.

I'm a little concerned that @nbauernfeind mentioned we might have liveness issues with Java applications. I assume Groovy or Python get executed as scripts using the session's `LivenessScope`. That is arguably out-of-scope for this PR, however.

File: server/src/main/java/io/deephaven/server/appmode/ApplicationServiceGrpcImpl.java (hunk 2 above)
Comment (devinrsmith) [reply server/src/main/java/io/deephaven/server/appmode/ApplicationServiceGrpcImpl.java]: This is actually a bug to not use the known type. Updated code is much cleaner.

File: server/src/main/java/io/deephaven/server/appmode/ApplicationServiceGrpcImpl.java (hunk 3 above)
Comment (devinrsmith) [reply server/src/main/java/io/deephaven/server/appmode/ApplicationServiceGrpcImpl.java]: It didn't work w/ calling propogateUpdates on thread. I could make the state management better for on-thread vs scheduler tracking, but I think it's a potentially unnecessary safety belt.

File: server/src/main/java/io/deephaven/server/appmode/ApplicationServiceGrpcImpl.java [hunk 4]
```diff
@@ -381,18 +251,159 @@ private void notifyObserverAborted() {
...
+
+    private void update(AppFieldId id, String description, String type) {
+        computeIfAbsent(id).update(description, type);
+    }
+
+    private void remove(AppFieldId id) {
+        computeIfAbsent(id).remove();
+    }
+
+    private State computeIfAbsent(AppFieldId id) {
```
Comment (rcaudy) [inline server/src/main/java/io/deephaven/server/appmode/ApplicationServiceGrpcImpl.java:266]: I would have called this something like `accumulate` or `accumulated`.

File: server/src/main/java/io/deephaven/server/appmode/ApplicationServiceGrpcImpl.java [hunk 5]
```diff
@@ -381,18 +251,159 @@ private void notifyObserverAborted() {
...
+            // If we send a remove, we'll be basing it off of our existing info
+            this.description = null;
+            this.type = null;
+        }
 
-    private static class LivenessTracker extends LivenessArtifact {
-        private <T> void maybeManage(T object) {
-            if (object instanceof LivenessReferent && DynamicNode.notDynamicOrIsRefreshing(object)) {
-                manage((LivenessReferent) object);
+        public void append(Updater builder) {
```
Comment (rcaudy) [inline server/src/main/java/io/deephaven/server/appmode/ApplicationServiceGrpcImpl.java:335]: Maybe better to rename `builder` to `updater`?

File: server/src/main/java/io/deephaven/server/appmode/ApplicationServiceGrpcImpl.java [hunk 6]
```diff
@@ -381,18 +251,159 @@ private void notifyObserverAborted() {
...
+                    .setTypedTicket(typedTicket(id, type))
+                    .setFieldName(id.fieldName)
+                    .setFieldDescription(description == null ? "" : description)
+                    .setApplicationId(id.applicationId())
+                    .setApplicationName(id.applicationName())
+                    .build();
+        }
+    }
+
+    // An "updater" as opposed to a "builder" - modifies state and is builder.
```
Comment (rcaudy) [inline server/src/main/java/io/deephaven/server/appmode/ApplicationServiceGrpcImpl.java:378]: Why not use JavaDoc format?

**Review by rcaudy — APPROVED** [review]
I made some strictly cosmetic comments. I like this change a lot. Merge it as-is, or make the small changes, either way it's good.

**Review by rcaudy — DISMISSED** [review]

**Review by nbauernfeind — APPROVED** [review]

File: engine/table/src/main/java/io/deephaven/stream/StreamToTableAdapter.java [hunk 7]
```diff
@@ -329,9 +344,12 @@ private void doRefresh() {
...
 
-        table.notifyListeners(new TableUpdateImpl(RowSetFactory.flat(newSize),
-                RowSetFactory.flat(oldSize), RowSetFactory.empty(),
-                RowSetShiftData.EMPTY, ModifiedColumnSet.EMPTY));
+        final QueryTable localTable = tableRef.get();
+        if (localTable != null) {
+            localTable.notifyListeners(new TableUpdateImpl(RowSetFactory.flat(newSize),
+                    RowSetFactory.flat(oldSize), RowSetFactory.empty(),
+                    RowSetShiftData.EMPTY, ModifiedColumnSet.EMPTY));
+        }
```
[the remark this answers is not in the document]
Comment (rcaudy) [reply engine/table/src/main/java/io/deephaven/stream/StreamToTableAdapter.java:352]: I'm ambivalent. We could, since we think that without the child table we have no reason to continue existing, but it's not strictly necessary. If `tableRef` has been cleared, the cleanup handler thread should drain its `RetainedReferenceTracker` from the reference queue, which should drop our reference count to 0. Probably better to be aggressive, though, in case someone has us in a `LivenessScope` keeping us active when there's no reason for it. I will make two "suggestion" comments.

File: engine/table/src/main/java/io/deephaven/stream/StreamToTableAdapter.java [hunk 8]
```diff
@@ -267,15 +272,23 @@ public void close() {
...
             doRefresh();
         } catch (Exception e) {
             log.error().append("Error refreshing ").append(StreamToTableAdapter.class.getSimpleName()).append('-')
                     .append(name).append(": ").append(e).endl();
-            table.notifyListenersOnError(e, null);
             updateSourceRegistrar.removeSource(this);
+            final QueryTable localTable = tableRef.get();
+            if (localTable != null) {
+                localTable.notifyListenersOnError(e, null);
+            }
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/stream/StreamToTableAdapter.java:291]:
```suggestion
            } else {
                close();
            }
```

File: engine/table/src/main/java/io/deephaven/stream/StreamToTableAdapter.java (hunk 7 above)
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/stream/StreamToTableAdapter.java:352]:
```suggestion
        } else {
            close();
        }
```

**Review by rcaudy — APPROVED** [review]
